### ISSUE TITLE
Remove the dynamic SE interface in 4.0

### ISSUE THREAD

**Issue by daverodgman** [Mbed-TLS/mbedtls#8151]
Remove the dynamic SE interface (supplanted by opaque drivers). `MBEDTLS_PSA_CRYPTO_SE_C` must be removed from `mbedtls_config.h`. All code related to this should be removed, except `mbedtls_psa_register_se_key()` which [first needs to change to work with new-style drivers](https://github.com/Mbed-TLS/mbedtls/issues/9255).

Mailing list thread: https://lists.trustedfirmware.org/archives/list/mbed-tls@lists.trustedfirmware.org/thread/7AM6WGPCIEHOZQ5G3M5LPZHNG5CSOPAP/

**Comment by mpg** [comment]
Note: a good portion of `docs/architecture/testing/driver-interface-test-strategy.md` can be removed as well.

**Comment by mpg** [comment]
Note: I think this will close https://github.com/Mbed-TLS/mbedtls/issues/3856 as well. Actually, when addressing this, we should search for "SE", `MBEDTLS_PSA_CRYPTO_SE_C`, etc. in open issues and close those that become irrelevant.

**Comment by gilles-peskine-arm** [comment]
> close those that become irrelevant.

Agreed that we should do that. But double check that the issues don't also apply to the new driver interface, or if they do that there is an issue for that.

I tend to call `MBEDTLS_PSA_CRYPTO_SE_C` the “old driver interface” or “dynamic secure element driver interface” or “dynamic SE drivers”.

**Comment by gilles-peskine-arm** [comment]
Note that at least `mbedtls_psa_register_se_key` [should stay but be adapted to the new kind of opaque drivers](https://github.com/Mbed-TLS/mbedtls/issues/9255).

**Comment by mimok** [comment]
I'm still using the "old" dynamic SE interface and i'm wondering whether the new PSA driver interface is actually in production or not before switching. I mean the driver interface cannot be fully defined through JSON files, you still need to manually modify the jinja templates. Is this really the way it is expected to work ?

**Comment by gilles-peskine-arm** [comment]
The C part is ready for production and actually used in production by several early adopters. Unfortunately, you do need to edit `psa_crypto_driver_wrappers*` manually at the moment. The way this is supposed to work is that you would provide a JSON file, but the script to generate the wrappers from JSON is not ready for production yet.

**Comment by DarkDust** [comment]
I was looking into how to wrap Apple keychain identities (certificate + private key) and APIs for use with mbedTLS (to do TLS 1.3). It seemed writing a dynamic PSA driver would be the way to go now, but then I found this issue. I've read through some of the docs and just want to know I'm understanding correctly what this means for me as a developer who wants to write a secure element driver:

* I have to write a driver, as described in the [docs](https://github.com/Mbed-TLS/mbedtls/blob/development/docs/psa-driver-example-and-guide.md).
* For the integration, I have to patch the mbedTLS source.
  * The files to patch should "just" be the ones in `scripts/data_files/{driver_jsons,driver_templates}/*`.
* Which means, with every update of mbedTLS I have to make sure my patch still applies and potentially fix any issue due to code changes in mbedTLS.

Is that correct? From what I'm understanding, maintaining a custom driver using `psa_register_se_driver` seems easier to implement and maintain. But according to this issue, the project's intention is to remove support for dynamic drivers in the long run, so right now it's already better to do it the way I've summarized above, right?

**Comment by gilles-peskine-arm** [comment]
That is correct, yes.

The old, dynamic interface only looks more stable because we've abandoned development on it. It's only partially implemented, and not tested much.

The new interface does require you to patch either the template files or the generated C files (I think it makes more sense to patch the generated C files) so that the dispatch code knows about your driver. This is temporary: we have a tool whose job is to read the template and your driver declaration, and generate the required dispatch code. Unfortunately, this tool is very incomplete, so it isn't ready for production.

You can use dynamic drivers in 2.28 and 3.6, but we won't help you unless you run into a major bug. In particular we won't add any features or resolve documented limitations. And we won't maintain that interface past 3.6.

If there is, specifically, demand for drivers that must be registered at runtime, we may add this ability back at some point after 4.1. It would be based on the new driver interface, not reusing any of the existing code of the old interface.

**Comment by mimok** [comment]
Hi @gilles-peskine-arm, i'm still wondering whether i should switch to the new driver interface.  Yesterday, i was digging into the code and it is unclear how an opaque SE driver could be actually implemented using the new driver interface. Indeed, a free slot ID is required to import a key into a SE however only the old driver interface allow to retrieve a free slot ID from this external SE. There is also no "atomic write" of the SE keys as implemented here: https://github.com/Mbed-TLS/mbedtls/blob/b6860cf7f9f4be0cc60f36909f6a5887008fb408/tf-psa-crypto/core/psa_crypto.c#L1975

**Comment by gilles-peskine-arm** [comment]
@mimok Beware that atomic write is not implemented for the old SE driver interface either! We have code that sets up a transaction for writing, but we don't have the code to recover an interrupted transaction.

The new SE driver interface all of the functionality of the old one, except for the ability to load drivers at runtime.

**Comment by mimok** [comment]
@gilles-peskine-arm 

> The new SE driver interface all of the functionality of the old one, except for the ability to load drivers at runtime.

My apologizes but i still don't understand how to use the new interface to implement a driver for an external secure element with storage using mbedTLS v3.6.2.

For example, when `MBEDTLS_PSA_CRYPTO_SE_C` is disabled there is no `slot_number` field in the attribute structure, no slot allocation mechanism, etc. I can't find in the source code what is described here: https://github.com/Mbed-TLS/mbedtls/blob/ce3c2dd30cf8c8d24d21aa6fe73d6642cbd4d494/docs/proposed/psa-driver-interface.md?plain=1#L1125

**Comment by gilles-peskine-arm** [comment]
With the old interface, using a secure element with storage is implemented at a proof-of-concept quality: you have the functionality, but if the device loses power at a bad time, you may end up with corrupted storage.

With the new interface, using a secure element with storage is not implemented yet.

That is filed in our [roadmap](https://mbed-tls.readthedocs.io/en/latest/project/roadmap/) as part of “[PSA Crypto][PSA Secure Element, Crypto Accelerator Support Enhancements]”. Note that it may be technically be possible to implement this in the 3.6 long-time support branch, but we (the Mbed TLS maintainer team) have no bandwidth to work on it until late 2025 at best.

**Comment by gilles-peskine-arm** [comment]
By the way, the current implementation can **use** keys that are stored in a secure element, by declaring them as built-in keys. See [`mbedtls_psa_platform_get_builtin_key()`](https://mbed-tls.readthedocs.io/projects/api/en/v3.6.2/api/group/group__psa__builtin__keys/) for more details. What we currently can't do is **create** keys that are stored in a secure elements, due to the lack of transaction handling.

**Comment by iabdalkader** [comment]
Is there a working replacement for the old API in mbedtls 3.6.4? More specifically, we need to perform ECDSA sign using keys stored on SE (similar to how `MBEDTLS_ECDSA_SIGN_ALT` used to work). Should we still use the deprecated dynamically registered SE API, or switch to the new one? If I understand correctly, from reading all relevant threads here, the new API does not yet support all operations. Is this true for 3.6.4? And does it at least support ECDSA sign without patching any files?

Note we're using Zephyr and we need to upstream this driver when it's done. I'm not sure how we can patch those files and upstream the changes to Zephyr.

Also, where can we find documentation for the new API? Are there any examples from Arm, mbedtls or third parties?

**Comment by gilles-peskine-arm** [comment]
The new API supports all the operations. What's missing is being able to plug in a driver without editing any upstream files. That's a work in progress.

There is a [driver example and guide](https://github.com/Mbed-TLS/TF-PSA-Crypto/blob/tf-psa-crypto-1.0.0/docs/psa-driver-example-and-guide.md), although unfortunately it's mostly focused on accelerators. The main documentation is the [driver interface reference](https://github.com/Mbed-TLS/TF-PSA-Crypto/blob/tf-psa-crypto-1.0.0/docs/proposed/psa-driver-interface.md).

**Comment by iabdalkader** [comment]
> What's missing is being able to plug in a driver without editing any upstream files. That's a work in progress.

So even if Zephyr updates to 4.0.0 soon, which is really unlikely, we still will not be able to add our drivers upstream. It sounds like we must use either the ALT option or the old API, at least for a while. Thanks for the information anyway, I really appreciate it!

**Comment by valeriosetti** [comment]
> Note we're using Zephyr and we need to upstream this driver when it's done. I'm not sure how we can patch those files and upstream the changes to Zephyr.

Currently in Zephyr if you want to support your driver through PSA interface, you need to modify `psa_driver_wrappers` similarly to the custom patches added by TF-M (see [this](https://github.com/zephyrproject-rtos/mbedtls/commit/f38d050d97fc59b347874ff8631e403b72fc57f8) and [this](https://github.com/zephyrproject-rtos/mbedtls/commit/0060e95618157d802093fa79b3dbe4de2534d2a8)).

> So even if Zephyr updates to 4.0.0 soon, which is really unlikely

Well, I guess it depends on when you need it, but the [4.0 transition is actually a work-in-progress](https://github.com/zephyrproject-rtos/zephyr/issues/97660). It's not that easy though because there are several occurrences to be replaced, but it's definitely happening.

**Comment by iabdalkader** [comment]
> Well, I guess it depends on when you need it, but the [4.0 transition is actually a work-in-progress](https://github.com/zephyrproject-rtos/zephyr/issues/97660). It's not that easy though because there are several occurrences to be replaced, but it's definitely happening.

If I understand correctly, even when that happens, we would still need to manually hard-code driver entry points into the PSA driver, similar to those commits you referenced, because auto generation is still a work in progress, so it's not in 4.0.0, correct?

To be clear: our goal is to just figure out if we can proceed with drivers for a couple of SE, and if so, what our current options are. We don't want to waste time on deprecated APIs that may not be accepted upstream. Thanks 🙏🏼

**Comment by valeriosetti** [comment]
> If I understand correctly, even when that happens, we would still need to manually hard-code driver entry points into the PSA driver, similar to those commits you referenced, because auto generation is still a work in progress, so it's not in 4.0.0, correct?

That's correct, manual patching is still required in 4.0. But there is work being discussed/planned on the Mbed TLS side (actually TF-PSA-Crypto to be more precise) to support drivers. If you want you be up to date on this you can join the [public Discord server](https://www.trustedfirmware.org/faq/) (psa-crypto-driver-api channel).
